draft-irtf-cfrg-hash-to-curve.md

```diff
@@ -356,869 +524,832 @@ document are to be interpreted as described in {{RFC2119}}.
-  computed in constant time, with respect to a fixed p, using
-  the equation x^((p-1)/2). For clarity, we will generally prefer using the
-  formula directly, and annotate the usage with this definition.
+The following procedure implements hash2base.
 
-- sqrt(x, p):
-  Computing square roots should be done in constant time where possible.
+~~~
+hash2base(x)
 
-  When p = 3 (mod 4), the square root can be computed as `sqrt(x, p) := x^(p+1)/4`.
-  This applies to P256, P384, and Curve448.
+Parameters:
+  1. H, a cryptographic hash function producing k bits.
+  2. F, a finite field F of order q and char(F)=p.
+Preconditions:  k >= floor(log2(p)) + 1.
+Input: x, an octet string to be hashed.
+Output: y, an element in F.
 
-  When p = 5 (mod 8), the square root can be computed by the following
-  algorithm, in which `sqrt(-1)` is a field element and can be precomputed.
-  This applies to Curve25519.
+Steps:
+  Case q=p
+    1. t1 = H(x)
+    2. t2 = OS2IP(t1)
+    3. y = t2 mod p
+    4. Output y
 
-~~~
-  sqrt(x, p) :=
-                 x^(p+3)/8     if x^(p+3)/4 == x
-      sqrt(-1) * x^(p+3)/8     otherwise
+  Case q=p^m  
+    1. t = H(x)
+    2. for i=0 to m-1
+    3.    t1 = H( t || I2OSP(i,2) )
+    4.    t2 = OS2IP(t1)
+    5.    e_i = t2 mod p
+    6. Output y = ( e_0, ..., e_{m-1} )
 ~~~
 
-  The above two conditions hold for most practically used curves, due to the
-  simplicity of the square root function. For others, a suitable constant-time
-  Tonelli-Shanks variant should be used as in {{Schoof85}}.
-
 # Deterministic Encodings  {#encodings}
 
 ## Interface
 
 The generic interface for deterministic encoding functions to elliptic curves is as follows:
 
 ~~~
-map2curve(alpha)
+(x, y) = map2curve(alpha)
 ~~~
 
-where alpha is a message to encode on a curve.
+where alpha is an octet string to be hashed to a point on the curve with affine
+coordinates (x, y) defined over F. Observe that each encoding requires that
+certain algebraic conditions must hold in order to be applied.
 
 ## Notation
 
-As a rough style guide for the following, we use (x, y) to be the output
-coordinates of the encoding method. Indexed values are used when the algorithm
-will choose between candidate values. For example, the SWU algorithm computes
-three candidates (x1, y1), (x2, y2), (x3, y3), from which the final (x, y)
-output is chosen via constant time comparison operations.
+As a rough style guide the following convention is used:
 
-We use u, v to denote the values in Fp output from hash2base, and use as
-initial values in the encoding.
+- All arithmetic operations are performed over the finite field, unless
+  otherwise be explicitly stated, e.g. integer arithmetic.
 
-We use t1, t2, ..., as reusable temporary variables. For notable variables, we
-will use a distinct name, for ease of debugging purposes when correlating with
-test vectors.
+- (x, y): are the coordinates of a point obtained by the encoding method.
```

Comment:
s/the coordinates/the affine coordinates/?

Reply:
Done